### PR TITLE
.Net: GoogleAI/VertexAI connector (mostly Gemini models)

### PR DESCRIPTION
**Pull request by Krzysztof318**
Description moved here #4680

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /// <inheritdoc />
+    public IAsyncEnumerable<StreamingTextContent> GetStreamingTextContentsAsync(
+        string prompt,
+        PromptExecutionSettings? executionSettings = null,
+        Kernel? kernel = null,
+        CancellationToken cancellationToken = default)
+    {
+        throw new NotImplementedException();
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs:119]: If there's no real streaming support, or none yet, at a minimum this can be implemented in terms of the non-streaming support, just yielding a single chunk.

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        if (textGenerationResponse is null)
+        {
+            throw new KernelException("Unexpected response from model")
+            {
+                Data = { { "ResponseData", body } },
+            };
+        }
+
+        return textGenerationResponse.Candidates.Select(c => new TextContent(c.Content.Parts[0].Text,
+            this.GetModelId(), textGenerationResponse)).ToList().AsReadOnly();
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs:76]: The AsReadOnly shouldn't be needed

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    private HttpRequestMessage VerifyArgumentsAndGetHTTPRequestMessage(string prompt, PromptExecutionSettings? executionSettings)
+    {
+        Verify.NotNullOrWhiteSpace(prompt);
+        if (executionSettings is not GeminiPromptExecutionSettings geminiExecutionSettings)
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs:82]: This should work with another type, including just the base. See what the OpenAI connector does. It tries to pull relevant info from the extra props dictionary, and everything else ends up just being a default.

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/TextGenerationResponse.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
+﻿#region HEADER
+// Copyright (c) Microsoft. All rights reserved.
+#endregion
+
+using System.Text.Json.Serialization;
+
+namespace Microsoft.SemanticKernel.Connectors.Gemini;
+
+internal class TextGenerationResponse
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Gemini/TextGeneration/TextGenerationResponse.cs:9]: Nit: sealed

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs (hunk 1 above)
Comment (Krzysztof318) [reply dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs:119]: Hi, thanks for your fast feedback :) For now this is a very initial pull request. I'll have more time to work on it over Christmas. Of course all comments from you are welcome.

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        httpRequestMessage.Headers.Add("User-Agent", HttpHeaderValues.UserAgent);
+        return httpRequestMessage;
+    }
+
+    private Uri GetUriForGetText() => new($"{GeminiApiEndpointBase}/{this._model}:generateContent?key={this._apiKey}");
+
+    private static ByteArrayContent GetHttpJsonContent(string prompt, GeminiPromptExecutionSettings geminiExecutionSettings)
+    {
+        var requestJsonObject = TextGenerationRequest.GenerateJsonFromPromptExecutionSettings(prompt, geminiExecutionSettings);
+        var requestUtf8Bytes = Encoding.UTF8.GetBytes(requestJsonObject.ToJsonString());
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs:106]: You should be able to serialize directly to bytes rather than going through an intermediary string

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/TextGenerationResponse.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+internal class TextGenerationResponseContent
+{
+    [JsonPropertyName("parts")]
+    public TextGenerationResponseParts[] Parts { get; set; }
+
+    [JsonPropertyName("role")]
+    public string Role { get; set; }
+}
+
+internal class TextGenerationResponseParts
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Gemini/TextGeneration/TextGenerationResponse.cs:42]: nit: sealed

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/TextGenerationResponse.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+internal class TextGenerationResponse
+{
+    [JsonPropertyName("candidates")]
+    public TextGenerationResponseCandidates[] Candidates { get; set; }
+
+    [JsonPropertyName("promptFeedback")]
+    public TextGenerationResponsePromptFeedback PromptFeedback { get; set; }
+}
+
+internal class TextGenerationResponseCandidates
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Gemini/TextGeneration/TextGenerationResponse.cs:18]: same for the rest

File: dotnet/src/Connectors/Connectors.Gemini/Connectors.Gemini.csproj [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    <Title>Semantic Kernel - Google Gemini AI connectors</Title>
+    <Description>Semantic Kernel connectors for Google Gemini. Contains clients for text generation and text embedding generation.</Description>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\SemanticKernel.Abstractions\SemanticKernel.Abstractions.csproj" />
+    <ProjectReference Include="..\..\SemanticKernel.Core\SemanticKernel.Core.csproj"/>
+  </ItemGroup>
+
+  <ItemGroup>
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Gemini/Connectors.Gemini.csproj:26]: I wouldn't expect this to be necessary

File: dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    private static Dictionary<string, object?> GetResponseMetadata(
+        GeminiResponse geminiResponse,
+        GeminiResponseCandidate candidate) => new()
+    {
+        ["FinishReason"] = candidate.FinishReason,
+        ["Index"] = candidate.Index,
+        ["TokenCount"] = candidate.TokenCount,
+        ["SafetyRatings"] = candidate.SafetyRatings?.Select(sr => new Dictionary<string, object?>
```
Comment (Krzysztof318) [inline dotnet/src/Connectors/Connectors.Gemini/TextGeneration/GeminiTextGenerationService.cs:91]: I was thinking here about create class like MetadataSafetyRatings but It will need casting when someone want to need access SafetyRatings metadata. What do you think @stephentoub ? I don't know do you allow to leak implementation details so used Dictionary.

**Comment by Krzysztof318** [comment]
> I think the connector should be called Vertex AI, as Gemini Pro is just an endpoint under the Vertex AI. Since we might also need to create an image generation endpoint which calls their [model](https://cloud.google.com/vertex-ai/docs/generative-ai/image/overview).
> 
> By the way, I'd be glad to assist.

I will consider it, curious to see what the SK crew has to say about it. 
Btw. i just implemented a gemini implementation based on google cloud (vertex ai)

**Comment by Krzysztof318** [comment]
I've stopped working on this brunch until review. Now I'm focused on gemini/functions branch.

**Comment by AshD** [comment]
Hi @Krzysztof318 

I pulled your branch and tried to build it but ran into the following compile errors. Maybe, I need to do some other stuff first.

_I upgraded the GoogleVertexAI connector to .NET8 and made it use the Semantic Kernel packages instead of project references._

`Error	CS0122	'JsonOptionsCache' is inaccessible due to its protection level	Connectors.GoogleVertexAI	`

Same error for other classes like HttpClientProvider, HttpHeaderValues, HttpRequest and Verify

Thanks,
Ash

**Comment by Krzysztof318** [comment]
> Hi @Krzysztof318
> 
> I pulled your branch and tried to build it but ran into the following compile errors. Maybe, I need to do some other stuff first.
> 
> _I upgraded the GoogleVertexAI connector to .NET8 and made it use the Semantic Kernel packages instead of project references._
> 
> `Error CS0122 'JsonOptionsCache' is inaccessible due to its protection level Connectors.GoogleVertexAI `
> 
> Same error for other classes like HttpClientProvider, HttpHeaderValues, HttpRequest and Verify
> 
> Thanks, Ash

Hi, I have not tested build with .net8 yet.  Hold off on upgrade to net8 until merge PRs #4421 and #4387. After merge this two PR I will upgrade project to net8 and c#12.
If you clone repo as is, it should build with no problem.

I'm not sure what did you do exactly but maybe you removed from Connectors.GoogleVertexAI.csproj this part:
```
<!-- IMPORT NUGET PACKAGE SHARED PROPERTIES -->
<Import Project="$(RepoRoot)/dotnet/nuget/nuget-package.props"/>
<Import Project="$(RepoRoot)/dotnet/src/InternalUtilities/src/InternalUtilities.props"/>
```
This would explain the error you got.

**Comment by AshD** [comment]
Thanks @Krzysztof318 for the quick answer. 

I got it to work :-)  after adjusting for the fact that Gemini does NOT accept system chat messages.

Great project. Thanks for building it. Off to look for an Amazon Bedrock connector.

Thanks,
Ash

**Comment by Krzysztof318** [comment]
@AshD 

> I got it to work :-) after adjusting for the fact that Gemini does NOT accept system chat messages.

I totally forgot about this, I added this to notes right now.
System message not supported exception is temporaly, I have idea for work-around this limitation in gemini.

**Comment by Krzysztof318** [comment]
@AshD If no one has started the amazon connector by then I can take care of it, but for now I am waiting for the new ADR to be created.

**Comment by AshD** [comment]
Hi @Krzysztof318  - I hacked a implementation for the Amazon connector using the Amazon AWS SDK. The code is pretty terrible but I can send it to you privately to use as a starter. The good news is that it works!

**Comment by Krzysztof318** [comment]
@AshD Thanks but I don't need it. I've had a preliminary look at the AWS SDK documentation and already have a rough outline of the project :)

File: dotnet/src/Connectors/Connectors.GoogleVertexAI.UnitTests/Connectors.GoogleVertexAI.UnitTests.csproj [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk"/>
+    <PackageReference Include="Moq"/>
+    <PackageReference Include="xunit"/>
+    <PackageReference Include="xunit.runner.visualstudio">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="System.Numerics.Tensors"/>
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.GoogleVertexAI.UnitTests/Connectors.GoogleVertexAI.UnitTests.csproj:30]: We should also include `coverlet.collector`, so this test project will be included in code coverage. 
```suggestion
    <PackageReference Include="coverlet.collector">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="System.Numerics.Tensors"/>
```

There is PR which already does that for all unit test projects: https://github.com/microsoft/semantic-kernel/pull/4610

File: dotnet/src/Connectors/Connectors.UnitTests/HuggingFace/TextGeneration/HuggingFaceTextGenerationTests.cs [hunk 11]
```diff
@@ -9,6 +9,7 @@
 using Microsoft.SemanticKernel;
 using Microsoft.SemanticKernel.Connectors.HuggingFace;
 using Microsoft.SemanticKernel.TextGeneration;
+using SemanticKernel.UnitTests;
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.UnitTests/HuggingFace/TextGeneration/HuggingFaceTextGenerationTests.cs:12]: Is this change required? Same for other files in `Connectors.UnitTests` project with the same change.

File: dotnet/src/Connectors/Connectors.GoogleVertexAI/Properties/AssemblyInfo.cs [hunk 12]
```diff
@@ -0,0 +1 @@
+﻿[assembly: System.Runtime.CompilerServices.InternalsVisibleTo("SemanticKernel.Connectors.UnitTests")]
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.GoogleVertexAI/Properties/AssemblyInfo.cs:1]: 1. Looks like test project should be `SemanticKernel.Connectors.GoogleVertexAI.UnitTests`.
2. I'm not sure this file is necessary, since you already made internals visible to test project in `Connectors.GoogleVertexAI.csproj` by using:
```
<ItemGroup>
  <InternalsVisibleTo Include="SemanticKernel.Connectors.GoogleVertexAI.UnitTests"/>
</ItemGroup>
```

File: dotnet/src/Connectors/Connectors.UnitTests/HuggingFace/TextGeneration/HuggingFaceTextGenerationTests.cs (hunk 11 above)
Comment (Krzysztof318) [reply dotnet/src/Connectors/Connectors.UnitTests/HuggingFace/TextGeneration/HuggingFaceTextGenerationTests.cs:12]: I'm sorry it's by accident, I'll correct it.

Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.UnitTests/HuggingFace/TextGeneration/HuggingFaceTextGenerationTests.cs:12]: No worries at all :)

Comment (Krzysztof318) [reply dotnet/src/Connectors/Connectors.UnitTests/HuggingFace/TextGeneration/HuggingFaceTextGenerationTests.cs:12]: @dmytrostruk I checked this, and yes, this using is required because I moved HttpMessageHandlerStub class from unit tests project to common directory. dotnet/src/InternalUtilities/test/HttpMessageHandlerStub.cs

Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.UnitTests/HuggingFace/TextGeneration/HuggingFaceTextGenerationTests.cs:12]: Makes sense, thanks for clarification!

**Comment by RogerBarreto** [comment]
@Krzysztof318 [Feature Branch](https://github.com/microsoft/semantic-kernel/tree/feature-connectors-gemini) was created to accommodate your great Gemini connector contribution, please, make small PRs as you progresses in the code changes and we can review it quickly.

Suggest adding bits by bits from this PR as we can review them separately.

Thanks

**Comment by Krzysztof318** [comment]
@RogerBarreto thanks :) For new commits it's great news but what with this PR? It's will difficult to split its now.

**Comment by Krzysztof318** [comment]
@RogerBarreto 
First split PR #4407

**Comment by Krzysztof318** [comment]
State of connector progress https://github.com/microsoft/semantic-kernel/issues/4680
First spit PR https://github.com/microsoft/semantic-kernel/pull/4407
